### ISSUE TITLE
Yield while building large results in Alternator

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#17995]
This issue is a reopened and updated version of https://github.com/scylladb/scylladb/issues/7926. That issue was prematurely closed, after only some of the subcases listed in it were solved (by https://github.com/scylladb/scylladb/pull/12351). Some of the subcases remained unsolved and recently caused stalls for users.

In commit https://github.com/scylladb/scylladb/commit/7953f7c65fcc3d41276a6d3602ec50a8ee40f0d0 we added yielding (i.e., preemption points) to the JSON parsing code in Alternator, so that a large request doesn't cause large stalls and high latency for other requests. However, there is also a similar issue for building large responses - such as a `GetItem` of a large item or a `Scan` or `Query` with a relatively large page. Every O(N) algorithm in such responses needs to be able to yield.

In in https://github.com/scylladb/scylladb/pull/12351 (UPDATE: and https://github.com/scylladb/scylladb/pull/18807) we fixed the following two cases:

* The low-level JSON printing code (rjson::print() called in server.cc) didn't yield.
 * For `BatchGetItem`, executor::batch_get_item() also ended with a loop over all items, which doesn't yield.

But after adding yields to the json-to-string function (rjson::print()), we still need to have yields in the code which generates this JSON in the first place. Specifically:

* In `Scan` or `Query` we call describe_items() which iterates over a page of items, converts them to JSON, filters them, etc. - currently without yielding during this loop. This loop can get fairly long when the data has tiny rows and a 1MB page can have a large number of them. Users have been seeing stalls in this case.
* The executor::describe_single_item() converts a single item to JSON, and needs to copy the item's entire content. In DynamoDB this is limited to 400 KB but in Scylla even this limit is not enforced and can be even larger (see https://github.com/scylladb/scylladb/issues/11880), so we should consider whether we should - or shouldn't - yield also inside this function. This describe_single_item() is used by `GetItem`, by `PutItem` to read the previous content (for read-modify-write operations and `ReturnValues`), and in `BatchGetItem` for each of the items.

Since we have coroutines now, writing yielding code has become easier, but some functions that today return an immediate value (specifically, the describe_items() and describe_single_item() functions) will need to be changed to return a future.

**Comment by nyh** [comment]
The following test reproduces the first case - a Query with a large page (the 1MB default) and very small rows, stalls while producing the 1MB result. To see a stall add `'--blocked-reactor-notify-ms', '5',` to test/alternator/run (the default setting is too high when running on my laptop).
```python
# The following test reproduces #17995: A Query returning a large page
# composed of many small rows, which causes a lot of processing work for
# outputting the result, and the risk (before #17995 is fixed) to stall.
# To see the stall, an option like '--blocked-reactor-notify-ms', '5'
# must be added to Scylla in test/alternator/run, as the default stall
# threshold is higher than the one that this test produces.
def test_query_large_page_small_rows(test_table_sn):
    p = random_string()
    # Experimentally, Scylla considers the rows we insert below (which each
    # have a 10-byte string partition key with a one byte name, a numeric
    # clustering key with a one byte name, and no other data) as being 32
    # bytes in size, and returns 32772 of these rows in one (nominally) 1MB
    # page of Query. So if the partition has just 30,000 rows, it will be
    # returned entirely in one page.
    N = 30_000
    with test_table_sn.batch_writer() as batch:
        for i in range(N):
            batch.put_item({'p': p, 'c': i})

    got_items = test_table_sn.query(KeyConditions={
        'p': {'AttributeValueList': [p], 'ComparisonOperator': 'EQ'}},
        ConsistentRead=True)['Items']
    n = len(got_items)
    assert n == N
```

On my laptop, running this consistently reports stalls of 14ms - 25ms.

**Comment by nyh** [comment]
I'm preparing now a patch to solve the `describe_items()` stall. The patch indeed avoid the largest (14ms-26ms) stall seen in the above test. However, there still remains a smaller 5ms stall, or perhaps several of them. The new culprit (or one of them?) appears to be the query page, perhaps `query_pager::handle_result()` (a non-future void function), which seems to handle the entire page - without yielding. I am guessing that this new stall is not Alternator specific, and could be reproduced also with CQL when reading a 1MB page from a partition with many tiny rows. If this is the case, we should consider opening a general non-Alternator-specific issue and closing this one.

The second case described above, `describe_single_item()` - with a very large single item - perhaps cannot be solved in the current state of Scylla. We already have a lot of code that works on whole items (CQL rows), and large single items (e.g., a large collection) will already cause stalls in various places, so I'm not sure we need to solve the Alternator-only case. But it won't hurt to have a test for this case as well.

**Comment by nyh** [comment]
> But after adding yields to the json-to-string function (rjson::print())

Sadly, as @avikivity showed in https://github.com/scylladb/scylladb/pull/18807, even this case was not actually fixed :-( The rjson::print() code still has a tight loop that generates all the writes of the JSON data, that doesn't yield. Later, it yields between those writes, but the original loop that goes over all the JSON content is still done without yielding. In #18807 Avi is planning to fix it by, unfortunately, using a seastar thread.

We really need to have more tests like test_query_large_page_small_rows (see above, and commit ba97fd98a38733c5a9a17c924615d25c281c98ca) that reproduces those stalls. That specific test isn't a regression test - it is relatively slow, and doesn't check anything (you only see stall warnings in the log if you configure it specially), but perhaps we can create such regression tests in a different test framework of slower tests, that runs Scylla with lower stall thresholds, and that looks at the logs to fail the test if any such warnings appeared. Or, manual testing also isn't that bad - it's better than what we did until now (basically, declare that some issue was fixed when it wasn't really fixed).

**Comment by bhalevy** [comment]
We've also seen a stall when destroying rjson::value on the error path, where it could help having a clear_gently method for it to destroy its structured contents in an async function the can yield.

**Comment by nyh** [comment]
> We've also seen a stall when destroying rjson::value on the error path, where it could help having a clear_gently method for it to destroy its structured contents in an async function the can yield.

Not just error path: In one user report I saw stalls in `alternator::make_streamed(rjson::value&& value)` - although this function makes sure to preempt during the the rjson::print call, at the very end the rjson::value object is no longer used so gets destroyed - and this destruction apparently happening without preemption.

So indeed make_streamed should call a new clear_gently on value. Or alternatively (but that will be harder), maybe we could somehow make rjson::print() take the moved value, and ask it to delete the object piece by piece while it is being printed.

This looks like a scenario which should be easy to reproduce in a test.

**Comment by bhalevy** [comment]
> > We've also seen a stall when destroying rjson::value on the error path, where it could help having a clear_gently method for it to destroy its structured contents in an async function the can yield.
> 
> Not just error path: In one user report I saw stalls in `alternator::make_streamed(rjson::value&& value)` - although this function - used by large Query/Scan responses or large BatchGetItem - makes sure to preempt during the the rjson::print call, at the very end the rjson::value object is no longer used so gets destroyed. This destruction apparently happening without preemption.
> 
> So indeed make_streamed should call a new clear_gently on value. Or alternatively (but that will be harder), maybe we could somehow make rjson::print() take the moved value, and ask it to delete the object piece by piece while it is being printed.

Yes, consuming the moved value gently sounds like a good idea.
That said, if there's an error, still the remainder in the value might be destroyed synchronously, so clear_gently would be our friend in that case as well.

> 
> This looks like a scenario which should be easy to reproduce in a test.

**Comment by nyh** [comment]
> So indeed make_streamed should call a new clear_gently on value. Or alternatively (but that will be harder), maybe we could somehow make rjson::print() take the moved value, and ask it to delete the object piece by piece while it is being printed.

@avikivity took the first approach in pull request https://github.com/scylladb/scylladb/pull/19968.

I believe it also applies to the error cases that @bhalevy mentioned, because make_streamed makes an effort to catch exceptions and rethrow them only after making sure to call close() and now - also the new destroy_gently(). 

After this patch, it's not clear what, if anything, is left of this issue. We should make an effort to write tests to find the remaining stalls and open specific issues about them - and close this one.

**Comment by avikivity** [comment]
The best solution is not to build a json document at all, and instead treat it as a serialization format and emit it incrementally.

**Comment by nyh** [comment]
> The best solution is not to build a json document at all, and instead treat it as a serialization format and emit it incrementally.

Definitely. Keeping everything JSON made some things in the Alternator query processing simpler (e.g., an expression can take the existing item and modify it or extract pieces of it, and this was easy to do with JSON), but as a first step, requests that return multiple items - namely Scan, Query and BatchGetItem, definitely don't need to generate a big JSON for everything and could work on small JSONs for individual items.

**Comment by mykaul** [comment]
@nyh - I think for the time being, we can close this issue - customer issue is mostly solved, even if we can improve it in the future, it's not on our roadmap, I believe.

**Comment by swasik** [comment]
> [@nyh](https://github.com/nyh) - I think for the time being, we can close this issue - customer issue is mostly solved, even if we can improve it in the future, it's not on our roadmap, I believe.

I think it well documents possible areas for improvement. Maybe we should just decrease priority and remove Field-Tier label? I will introduce such changes.
